File: androidApp/src/main/java/com/example/stories/android/ui/communityStories/CommunityStoriesListViewModel.kt
package com.example.stories.android.ui.communityStories

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.example.stories.viewModel.CommunityStoriesListCommonViewModel

class CommunityStoriesListViewModel(userId: String) : ViewModel() {

    private val commonViewModel = CommunityStoriesListCommonViewModel(userId = userId, coroutineScope = viewModelScope)

    val userAndStoriesLoadStatus get() = commonViewModel.userAndStoriesLoadStatus

    fun refreshData() = commonViewModel.refreshData()

    class Factory(private val userId: String) : ViewModelProvider.NewInstanceFactory() {
        override fun <T : ViewModel> create(modelClass: Class<T>): T = CommunityStoriesListViewModel(userId) as T
    }
}

File: iosApp/iosApp/ui/communityStories/CommunityStoriesListViewModel.swift
import Foundation
import shared

extension CommunityStoriesListScreen {
    @MainActor class CommunityStoriesListViewModel : ObservableObject, ViewLifeCycleObserver {
        
        private let userId: String
        private var commonViewModel: CommunityStoriesListCommonViewModel?
        
        init(userId: String) {
            self.userId = userId
        }
        
        @Published var userAndStoriesLoadStatus: LoadStatus<Reference<(User, [History])>>? = nil
        
        func refreshData(showLoading: Bool) {
            commonViewModel?.refreshData(showLoading: showLoading)
        }
        
        func startObserving() {
            let commonViewModel = CommunityStoriesListCommonViewModel(userId: userId)
            self.commonViewModel = commonViewModel
            commonViewModel.userAndStoriesLoadStatus.subscribe(scope: commonViewModel.viewModelScope) { userAndStoriesLoadStatus in
                if let userAndStoriesLoadStatus {
                    self.userAndStoriesLoadStatus = userAndStoriesLoadStatus.mapData { userAndStories in
                        let user = userAndStories.first!
                        let stories = userAndStories.second as? [History] ?? []
                        return Reference(value: (user, stories))
                    } as? LoadStatus<Reference<(User, [History])>>
                }
            }
        }
        
        func stopObserving() {
            commonViewModel?.dispose()
        }
    }
}
